### PR TITLE
Bookmarks: Undo behavior

### PR DESCRIPTION
**Pull request by malmstein**
Task/Issue URL: https://app.asana.com/0/1174433894299346/1205171041444238

### Description
This PR changes the delete behavior for saved sites and folders. Instead of deleting / inserting when undo, we don’t touch the repository until the snackbar has been hidden or the user leaves the activity whilst the snackbar is shown.

### Steps to test this PR

_Bookmark delete_
- [ ] Add a bookmark and open the Bookmarks screen
- [ ] Delete it
- [ ] Verify snackbar shows and bookmark is not in the list
- [ ] Wait for the snackbar to disappear
- [ ] Leave the screen and come back again
- [ ] Bookmark should have been deleted

_Bookmark undo delete_
- [ ] Add a bookmark and open the Bookmarks screen
- [ ] Delete it
- [ ] Verify snackbar shows and bookmark is not in the list
- [ ] Press Undo
- [ ] Verify bookmark is shown again
- [ ] Leave the screen and come back again
- [ ] Bookmark should be in the list

Repeat the same for Favourites and Folders

_Bookmark delete (Sync enabled)_
- [ ] Enable Sync
- [ ] Add a bookmark and open the Bookmarks screen
- [ ] Verify Sync operations starts
- [ ] Delete it
- [ ] Verify Sync operation does not start
- [ ] Verify snackbar shows and bookmark is not in the list
- [ ] Wait for the snackbar to disappear
- [ ] Verify Sync operation starts

_Bookmark undo delete (Sync enabled)_
- [ ] Enable Sync
- [ ] Add a bookmark and open the Bookmarks screen
- [ ] Verify Sync operations starts
- [ ] Delete it
- [ ] Verify Sync operation does not start
- [ ] Verify snackbar shows and bookmark is not in the list
- [ ] Press Undo
- [ ] Verify bookmark is shown again
- [ ] Verify Sync operation does not start

### CODE REVIEW

**Comment by malmstein** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3507** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3507" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3507?utm_source=stack-comment).

**Review by cmonfortep — COMMENTED** [review]
Tested this first and found a few bugs, posted info in the internal task.

Will hold review until we fix that.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt [hunk 1]
```diff
@@ -142,19 +138,89 @@ class BookmarksViewModel @Inject constructor(
...
 
-    private fun delete(savedSite: SavedSite) {
+    private fun hide(savedSite: SavedSite) {
+        // we don't delete, just remove from the ui list
+        when (savedSite) {
+            is Bookmark -> {
+                val bookmarks = viewState.value?.bookmarks!!.toMutableList()
+                val index = bookmarks.indexOf(savedSite)
+                bookmarks[index] = savedSite.copy(deleted = "1")
+                val bookmarkFolders = viewState.value?.bookmarkFolders!!
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt:151]: I see too many `!!` in the new logic (this method and the other ones), which is kind of dangerous and I think can be the root cause of one bug I've reported. can we apply early return or use `let` instead?

**Review by cmonfortep — COMMENTED** [review]
I've done another pass here, and sorry to say that I've found again a few bugs. Will post them in the internal task so I can attach videos and steps to repro.

Left some comments as well on where the bugs maybe coming from, hope that helps.

Did a review on the code, and overall the approach looks fine. However I've noticed we are just changing the undo logic in the bookmarks screen, but missing other scenarios like:
- Favorites quick access grid appears in 3 places: new tab, autocomplete, system search (screen that widget opens) -> on those ones, the new undo logic is not present
- Edit Bookmark also allow the user to delete bookmark. You can do that action from browser tab (menu -> edit boomark -> delete), or from bookmarks screen (edit bookmark -> delete)

We should also add them here.

**Comment by malmstein** [comment]
closing this for now, won’t work on it for a few days

**Pull request closed without merging**